### ISSUE TITLE
Passing arguments to Jupyter notebooks differs between directly passed and function-generated arguments

### ISSUE THREAD

**Issue by sebhoerl** [snakemake/snakemake#2753]
**Snakemake version**
8.6.0

**Describe the bug**
I tried to pass, analogously to the input files, a named dictionary as parameters to a Jupyter notebook:

```python
def params(wildcards):
  return { "a": "abc", "b": "uvw" }

rule:
  params: unpack(params)
  notebook: "mynotebook.ipynb"
```

Apparently, this is not the right synax, snakemake complains with the following error:

```
unpack() is not allowed with params. Simply return a dictionary which can be directly .used, e.g. via {params[mykey]}.
```

Removing the `unpack` wrapper runs, but `snakemake.params` is structured as follows in the notebook:

```python
[ { "a": "abc", "b": "uvw" } ]
```

So I need to access *a* by calling `snakemake.params[0]["abc"]`, i.e., the parameters seem to be passed as a list (different to what is indicated in the error message).

However, I would expect the set-up from above do the same as:
```python
rule:
  params: a="abc", b="uvw"
  notebook: "mynotebook.ipynb"
```

In this case, the parameters are *not* passed as a list, but directly as a dictionary and I can directly access `params["a"]`.

The bottom-line is that I think that with passing a function, we should have the same behaviour as when passing the parameters directly inside the rule. Otherwise, the notebook will not be universally useable. Alternatively, like for the `input` slot, `unpack` should be allowed here.

**Logs**

**Minimal example**

See above.

**Additional context**

**Comment by sebhoerl** [comment]
Quick fix if you want to make your notebook robust against the way parameters are passed:

```python
params = snakemake.params[0] if len(snakemake.params) == 1 and len(snakemake.params.keys()) == 0 else snakemake.params
```

**Comment by durr1602** [comment]
Same here with snakemake 8.18.0

I noticed that if you're calling a script instead of a notebook, the output does not seem to be interpreted the same way. Here is part of my Snakefile:

`rule stats:
    input:
        cutadapt_logs = expand(rules.cutadapt.log, sample=S),
        pandaseq_logs = expand(rules.pandaseq.log, sample=S),
        vsearch_logs = expand(rules.vsearch.log, sample=S)
    output:
        'read_stats.csv'
    resources:
        threads = 1,
        time = "00:01:00"
    conda:
        'envs/jupyter_basic.yaml'
    script:
        'scripts/get_read_stats.py'

rule parse_fasta:
    input:
        fasta_files = expand(rules.vsearch.output, sample=S),
        read_stats = rules.stats.output[0],
        expected_mutants = rules.generate_mutants.output[0]
    output:
        unexp_rc_plot = 'graphs/unexp_rc_plot.svg'
    resources:
        mem_gb = 2,
        threads = 1,
        time = "00:01:00"
    log:
        notebook="logs/notebooks/parse_fasta.ipynb"
    conda:
        'envs/jupyter_plotting.yaml'
    notebook:
        'template_notebooks/parse_fasta.py.ipynb'
`
In get_read_stats_py, I call a function with: `generate_read_stats(snakemake.input.cutadapt_logs, snakemake.input.pandaseq_logs, snakemake.input.vsearch_logs, snakemake.output)`

So in summary, output seems to be interpreted as a string when calling a script, or as a list of string when calling a notebook

**Comment by durr1602** [comment]
Edit related to my previous comment. I realized the script was never actually outputting using the specified `snakemake.output` variable. When I corrected, it did not work unless I wrote `snakemake.output[0]`. So... my bad!
